editor/groups_editor.cpp

```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+	ClassDB::bind_method("_item_edited", &GroupsEditor::_item_edited);
+
+	ClassDB::bind_method("_add_scene_group", &GroupsEditor::_add_scene_group);
+	ClassDB::bind_method("_rename_scene_group", &GroupsEditor::_rename_scene_group);
+	ClassDB::bind_method("_remove_scene_group", &GroupsEditor::_remove_scene_group);
+	ClassDB::bind_method("_set_group_checked", &GroupsEditor::_set_group_checked);
+
+	ClassDB::bind_method("_remove_node_references", &GroupsEditor::_remove_node_references);
+	ClassDB::bind_method("_rename_node_references", &GroupsEditor::_rename_node_references);
 
-void GroupsEditor::_show_group_dialog() {
-	group_dialog->edit();
+	ClassDB::bind_method("_cache_scene_groups", &GroupsEditor::_cache_scene_groups);
 }
 
-void GroupsEditor::_bind_methods() {
-	ClassDB::bind_method("update_tree", &GroupsEditor::update_tree);
-	ClassDB::bind_method("_group_selected", &GroupsEditor::_group_selected);
+void GroupsEditor::_node_removed(Node *p_node) {
+	if (scene_root_node == p_node) {
+		scene_groups_for_caching = scene_groups;
+		call_deferred(SNAME("_cache_scene_groups"), p_node);
+		scene_root_node = nullptr;
+	}
+
+	if (scene_root_node && scene_root_node == p_node->get_owner()) {
+		_global_group_changed();
+	}
 }
 
 GroupsEditor::GroupsEditor() {
 	node = nullptr;
+	scene_tree = SceneTree::get_singleton();
 
-	VBoxContainer *vbc = this;
-
-	group_dialog = memnew(GroupDialog);
-
-	add_child(group_dialog);
-	group_dialog->connect("group_edited", callable_mp(this, &GroupsEditor::update_tree));
-
-	Button *group_dialog_button = memnew(Button);
-	group_dialog_button->set_text(TTR("Manage Groups"));
-	vbc->add_child(group_dialog_button);
-	group_dialog_button->connect("pressed", callable_mp(this, &GroupsEditor::_show_group_dialog));
+	ED_SHORTCUT("groups_editor/delete", TTR("Delete"), Key::KEY_DELETE);
+	ED_SHORTCUT("groups_editor/rename", TTR("Rename"), Key::F2);
+	ED_SHORTCUT_OVERRIDE("groups_editor/rename", "macos", Key::ENTER);
 
 	HBoxContainer *hbc = memnew(HBoxContainer);
-	vbc->add_child(hbc);
-
-	group_name = memnew(LineEdit);
-	group_name->set_h_size_flags(Control::SIZE_EXPAND_FILL);
-	hbc->add_child(group_name);
-	group_name->connect("text_submitted", callable_mp(this, &GroupsEditor::_add_group));
-	group_name->connect("text_changed", callable_mp(this, &GroupsEditor::_group_name_changed));
+	add_child(hbc);
 
 	add = memnew(Button);
-	add->set_text(TTR("Add"));
+	add->set_flat(true);
+	add->set_tooltip_text(TTR("Add a new group."));
+	add->connect("pressed", callable_mp(this, &GroupsEditor::_show_add_group_dialog));
 	hbc->add_child(add);
-	add->connect("pressed", callable_mp(this, &GroupsEditor::_add_group).bind(String()));
+
+	filter = memnew(LineEdit);
+	filter->set_clear_button_enabled(true);
+	filter->set_placeholder(TTR("Filter Groups"));
+	filter->set_h_size_flags(Control::SIZE_EXPAND_FILL);
+	filter->connect("text_changed", callable_mp(this, &GroupsEditor::_filter_changed));
```

Comment:
```suggestion
	filter->connect("text_changed", callable_mp(this, &GroupsEditor::_update_tree).unbind(1));
```
`_filter_changed()` can be removed.